R/calculate-grades.R

```diff
@@ -1,261 +1,441 @@
+#' @param policy_item An item from policy file
+#'
+#' @return A data frame
+#'
+#' 
+#' @export
+
+get_category_grade <- function(grades_mat, policy_item){
+  #get all keys except category and assignments (which are not functions)
+  keys <- names(policy_item)[-which(names(policy_item) %in% c("category", "assignments", "weights"))]
+  #all assignments + their associated cols for this category
+  for (key in keys){
+    grades_mat <- get(key)(grades_mat, policy_item[[key]], policy_item$category, 
+                           policy_item$assignments, policy_item$weights)
+  }
+  
+  ## insert code to re-add new/editted columns from grades_mat into grades_mat
+  return (grades_mat)
+}
+
+
+#' Key Functions
+#'
+#' @description
+#' 
+#' A collection of functions to refer to correct classification of grading functions using policy file keys.
+#'
+#' * `score()` computes the percentage score and saves back into raw points col
+#' 
+#' * `aggregation()` computes score for category
+#'
+#' * `lateness()` applies any relevant lateness penalty.
+#' 
+#' * `drop_n_lowest()` drops n lowest assignment scores; if n >= num of assignments, return top assignment
+#' 
+#' * `aggregation_max_pts()` computes max points for category.
+#' 
+#' * `aggregation_lateness()` computes max points for category.
+#'
+#' @param grades_mat Matrix with assignments + associated cols for that category
+#' @param policy_line Policy list item for that key
+#' @param category Category name
+#' @param assignments Assignment names for this category
+#' @param weights Weights for `weighted_mean`
+#' 
+#' @return A matrix
+#'
+#' @family {Key Functions}
+#' 
+#' @export
+score <- function(grades_mat, policy_line, category, assignments, weights = c()){
+  get(policy_line)(grades_mat, assignments)
+}
+
+#' @rdname score
+#' @export
+aggregation <- function(grades_mat, policy_line, category, assignments, weights = c()){
+  get(policy_line)(grades_mat, category, assignments, weights)
+}
+
+#' @rdname score
+#' @export
+lateness <- function(grades_mat, policy_line, category, assignments, weights = c()){
+  original_late_mat <- grades_mat[, paste0(assignments, " - Lateness (H:M:S)")]
+  for (late_policy in policy_line){
+   grades_mat <- get(names(late_policy))(grades_mat, late_policy, original_late_mat, assignments)
+  }
+  return (grades_mat)
+}
+
+#' @rdname score
+#' @export
+drop_n_lowest <- function(grades_mat, policy_line, category, assignments, weights = c()){
+  n <- policy_line
+  grades_mat[, assignments] <- t(apply(t(grades_mat[, assignments]), 2, function(assign){
+    if (length(assignments) == 1) return (assign)
+    lowest_indices <- order(assign)[1:n]
+    if (n >= length(assignments)) {
+      lowest_indices <- order(assign)[1:(length(assignments)-1)]
+    }
+    assign[lowest_indices] <- NA
+    return(assign)
+  }))
+  return (grades_mat)
+}
+
+#' @rdname score
+#' @export
+aggregation_max_pts <- function(grades_mat, policy_line, category, assignments, weights = c()){
```

Comment:
This brings up an interesting idea. Right now a policy file (I think) would look like this:

```yaml
...
aggregation: equally_weighted
aggregation_max_pts: sum
```
Another representation would be:

```yaml
...
aggregation:
  score: equally_weighted
  max_pts: sum
```
What do you think?

Reply:
Hmm, there's also aggregation of lateness, and `score` is already a function for how to calculate the percentage of an assignment (e.g. raw_over_max)... that's one of the limitations of the "meta" approach: you can't have repeat names for functions. Also, this assumes that all aggregation is done at the same step (not necessarily an issue, but just something to think through, so I'm definitely up for the discussion and can address in a separate pull request later)